_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,637 @@
+
+### Add the Auth0 Identity Provider
+
+Navigate to **http://keycloak.\<namespace\>.\<public-ip\>.nip.io**, sign in with admin/admin. On the welcome page, choose **Administration Console**. On the left menu, for the top options,  choose the **jhipster** realm. Then, at the bottom of the menu, choose **Identity providers**. In the User-defined section, choose **OpenID Connect v1.0**. Fill in the provider configuration as follows:
+
+- RedirectURI: **(pre-filled)**
+- Alias: **auth0**
+- User discovery endpoint: **On**
+- Discovery endpoint: **https://\<auth0-tenant\>/.well-known/openid-configuration**
+- Client authentication: **Client secret sent as post**
+- Client ID: **auht0-client-id**
+- Client Secret: **auth0-client-secret**
+
+Click on **Add** to continue the configuration. Below the Client Secret field, click on **Advanced**. In the _Scopes_ field, set `openid profile email offline_access` and click on **Save**.
+
+On the left menu, choose **Authentication**. In the flows table, choose **browser**. In the _Identity Provider Redirector_ step, click the **gear** icon. Set an alias for the configuration, and set **auth0** as the default identity provider. This configuration will skip the Keycloak sign-in form, and display the Auth0 sign-in form directly.
+
+{% img blog/keycloak-kubernetes-prod/auth0-redirector.png alt:"Identity provider redirector to Auth0" width:"450" %}{: .center-image }
+
+Go back to **Authentication**, and choose the **first broker login** flow. Disable the _Review Profile_ step.
+
+### Map the roles claim
+
+The users that sign in through Auth0 are imported to Keycloak. The role assigned in Auth0 must be mapped to the pre-configured roles in Keycloak for the authorization mechanism to work. In the Keycloak Administration Console, go to **Identity providers** again, and choose **auth0**. Then choose the **Mappers** tab. Click **Add mapper**. Fill in the values as follows and click **Save**:
+
+- Name: **ROLE_USER**
+- Sync mode override: **Force** (update the user roles during every login)
+- Mapper type: **Claim to Role**
+- Claim: **https://www\\.jhipster\\.tech/roles** (escape the dot (.) with backslash (\\.))
+- Claim Value: **ROLE_USER**
+- Role: Select **ROLE_USER**
+
+Repeat the process for mapping `ROLE_ADMIN`.
+
+{% img blog/keycloak-kubernetes-prod/role-mapper.png alt:"Add mapper form" width:"800" %}{: .center-image }
+
+### Test Keycloak as an identity broker
+
+Navigate to **http://gateway.\<namespace\>.\<public-ip\>.nip.io**, the gateway home must display. Click on **sign in** and the browser should be redirected to the Auth0 sign-in form:
+
+{% img blog/keycloak-kubernetes-prod/auth0-login.png alt:"Add mapper form" width:"350" %}{: .center-image }
+
+After signing in, the gateway home page will display the username (email). If roles were mapped correctly, the user should be able to create entities (granted by ROLE_USER), and the Administration menu will also display (granted by ROLE_ADMIN).
```

Comment:
I'm not sure if we should surround ROLE_USER and ROLE_ADMIN with back tics. It looks fine without.